### PR TITLE
project20-Vlad-Ati

### PR DESCRIPTION
**Pull request by takrami**
Link to my deployed project:
https://elastic-hawking-e24322.netlify.app

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,11 @@
...
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+Main problem was connection to cloud MongoDB database
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+The application is alive on Netlify server: https://elastic-hawking-e24322.netlify.app/
```
Comment (KarolCislowski) [inline README.md:11]: To fix it you need to create `_redirects` file in the `public` directory of your react project and paste it inside:
``` 
/* /index.html 200
```

image

File: backend/server.js [hunk 2]
```diff
@@ -1,30 +1,130 @@
...
+    unique: true,
+  },
+  password: {
+    type: String,
+    required: true,
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString("hex"),
+  },
```
Comment (KarolCislowski) [inline backend/server.js:42]: Nice model 👍🏻

File: frontend/src/components/Header.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  const onButtonClick = () => {
+    batch(() => {
+      dispatch(user.actions.setUsername(null))
+      dispatch(user.actions.setAccessToken(null))
+      dispatch(notes.actions.setNotes([]))
+
+      localStorage.removeItem('user')
+    })
+  }
```
Comment (KarolCislowski) [inline frontend/src/components/Header.js:23]: great that you remembered to use the localStorage to save your user's data

File: frontend/src/components/Login.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        if (data.success) {
+          batch(() => {
+            dispatch(user.actions.setUsername(data.username))
+            dispatch(user.actions.setAccessToken(data.accessToken))
+            dispatch(user.actions.setErrors(null))
+
+            localStorage.setItem('user', JSON.stringify({
+              username: data.username,
+              accessToken: data.accessToken
+            }))
```
Comment (KarolCislowski) [inline frontend/src/components/Login.js:54]: 💯

File: frontend/src/components/Main.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        method: "GET",
+        headers: {
+          Authorization: accessToken,
+        },
+      };
+      fetch(API_URL("notes"), options)
+        .then((res) => res.json())
+        .then((data) => dispatch(notes.actions.setNotes(data)));
+    }
+  }, [accessToken, dispatch]);
```
Comment (KarolCislowski) [inline frontend/src/components/Main.js:39]: Why 2 effects if you can have just one :

```jsx 
useEffect(() => {
    if (!accessToken) {
      history.push('/login')
    } else {
      const options = { ...}
      ...
    }
  }, [accessToken, history, dispatch])
  ```

File: frontend/src/reducers/user.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    setAccessToken: (store, action) => {
+      store.accessToken = action.payload
+    },
+    setErrors: (store, action) => {
+      store.errors = action.payload
+    }
+  }
+})
+
+export default user
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:34]: great reducer!

**Review by KarolCislowski — APPROVED** [review]
It is really great app. I will approve it now but please fix the Netlify error page issue 😸

File: README.md (hunk 1 above)
Comment (takrami) [reply README.md:11]: Oh thank you so much for your help :)

**Pull request closed without merging**